myia/compile/backends/python/code_generator.py

```diff
@@ -0,0 +1,286 @@
+
+
+def str_getattr(c, *params):
+    """Formatter for apply getattr."""
+    assert len(params) == 2
+    obj, symbol = params
+    return f"{c.label(obj)}.{c.label(symbol)}"
+
+
+SIMPLE_MAP = {
+    operator.add: "{} + {}",
+    operator.and_: "{} & {}",
+    operator.contains: "{} in {}",
+    operator.eq: "{} == {}",
+    operator.floordiv: "{} // {}",
+    operator.ge: "{} >= {}",
+    operator.getitem: "{}[{}]",
+    operator.gt: "{} > {}",
+    operator.invert: "~{}",
+    operator.is_: "{} is {}",
+    operator.is_not: "{} is not {}",
+    operator.le: "{} <= {}",
+    operator.lshift: "{} << {}",
+    operator.lt: "{} < {}",
+    operator.mod: "{} % {}",
+    operator.mul: "{} * {}",
+    operator.ne: "{} != {}",
+    operator.neg: "-{}",
+    operator.not_: "not {}",
+    operator.or_: "{} | {}",
+    operator.pos: "+{}",
+    operator.pow: "{} ** {}",
+    operator.rshift: "{} >> {}",
+    operator.sub: "{} - {}",
+    operator.truediv: "{} / {}",
+    operator.truth: "bool({})",
+    operator.xor: "{} ^ {}",
+}
+COMPLEX_MAP = {
+    "user_switch": str_user_switch,
+    "switch": str_user_switch,
+    "make_tuple": str_make_tuple,
+    "make_list": str_make_list,
+    "make_dict": str_make_dict,
+    "apply": str_apply,
+    getattr: str_getattr,
+}
+
+
+class NodeLabeler:
+    """Node labeler used in Python backend.
+
+    Combine a Labeler (from myia.utils.info) and a node cache to generate default names when necessary.
+    """
+
+    def __init__(self):
+        """Initialize."""
+        self.cache = {}
+        self.default_name_counter = Counter()
+        self.lbl = Labeler(
+            relation_translator=self._relation_translator,
+            name_generator=self._name_generator,
+            disambiguator=self._disambiguator,
+            object_describer=self._object_describer,
+        )
+
+    def __call__(self, node):
+        """Generate label for given node."""
+        if isinstance(node, Node) and node.is_constant_graph():
+            # Use labeler for graph in constant node.
+            return self.lbl(node.value)
+        else:
+            # Use labeler for anything else.
+            return self.lbl(node)
+
+    @classmethod
+    def _relation_translator(cls, rel):
+        return f"{rel}_"
+
+    @classmethod
+    def _name_generator(cls, identifier):
+        return f"_apply{identifier}"
+
+    @classmethod
+    def _disambiguator(cls, label, id):
+        return f"_{label}_{id}"
+
+    def _object_describer(self, node):
+        if (
+            isinstance(node, Node)
+            and node.is_constant()
+            and not node.is_constant_graph()
+        ):
+            return str(node.value)
+
+        if node.debug:
+            return None
+
+        if node not in self.cache:
+            name = f"_{type(node).__name__.lower()}"
+            self.default_name_counter.update([name])
+            self.cache[node] = f"{name}{self.default_name_counter[name]}"
+        return self.cache[node]
```

Comment:
Hi! Indeed, it will work. but it will use pattern `_apply{id}` everywhere (because of `name_generator`), both for function names, apply names, and parameter names. Here I try to generate different names for each kind of node, to keep a bit of information even if debug info is not available. So, functions are named `_graph{id}`, applies are named `_apply{id}`, and parameters are named `_parameter{id}`.

Should I still remove it and let labeler generate default names ?

Reply:
Ok! So I will just return None.